presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -620,6 +624,12 @@ private static boolean isScheduled(Optional<StageInfo> rootStage)
 
     public boolean setFinalQueryInfo(QueryInfo currentInfo, QueryInfo newInfo)
     {
+        queryInfoState.set(FINAL_INFO_AVAILABLE);
         return finalQueryInfo.compareAndSet(currentInfo, newInfo);
     }
+
+    public void addFinalQueryInfoListener(StateChangeListener<QueryInfoState> queryInfoListener)
```

Comment:
I think we should do something like this:
```java
    private final StateMachine<Optional<QueryInfo>> finalQueryInfo = new StateMachine<>("finalQueryInfo", executor, Optional.empty());

    public void addStateChangeListener(StateChangeListener<QueryInfo> stateChangeListener)
    {
        AtomicBoolean done = new AtomicBoolean();
        finalQueryInfo.addStateChangeListener(finalQueryInfo -> {
            if (finalQueryInfo.isPresent() && done.compareAndSet(false, true))  {
                stateChangeListener.stateChanged(finalQueryInfo.get());
            }
        });
    }
```

With this we don't need the `enum` and only have one state variable.